### PR TITLE
fix: Restructure and update role-invite flow 

### LINKED ISSUE #2771: 400 being encountered after accepting a role-invite
**Describe the bug**

When a user accepts a role-invite, he/she encounters a 400 error.

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to Role-invite while the user is logged in
2. Enter the invite token
3. See error

**Expected behaviour**

It should've worked without any error

**GIF**

ezgif com-video-to-gif (4)

**Desktop (please complete the following information):**
 - OS: [e.g. iOS] Ubuntu 16.04
 - Browser [e.g. chrome, safari] Chrome

**Additional context**

Working on it

### PR DESCRIPTION
**Pull request by uds5501**
Fixes #2771 

#### Changes proposed in this pull request:

- Fix transition syntax after the refactor

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by uds5501** [comment]
Server PR : fossasia/open-event-server#6015

**Comment by uds5501** [comment]
@kushthedude @shreyanshdwivedi @mrsaicharan1 Please review!

**Comment by uds5501** [comment]
@CosmicCoder96 @niranjan94 After this change, I think the element of network requests is resolved.
If I need to do something else too, please let me know.

**Comment by uds5501** [comment]
@mrsaicharan1 @shreyanshdwivedi @kushthedude Review please

**Comment by niranjan94** [comment]
@shreyanshdwivedi @kushthedude please review the code before pressing approve, other-wise no point in having peer-reviews. :)

**Comment by kushthedude** [comment]
@uds5501 Can you make requested changes ASAP

**Comment by uds5501** [comment]
@niranjan94 please have a look now

File: app/routes/public/role-invites.js [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
     };
-    this.loader
+    await this.loader
       .post('/role_invites/user', payload)
       .then(user => {
-        if (this.get('session.isAuthenticated')) {
-          if (this.get('authManager.currentUser.email') === user.email) {
+        if (this.session.isAuthenticated) {
+          if (this.authManager.currentUser.email === user.email) {
             this.loader
```
Comment (uds5501) [inline app/routes/public/role-invites.js]: @niranjan94 adding an `await` here threw a build error so removed it, will it be any problem for now?

**Comment by shreyanshdwivedi** [comment]
@niranjan94 actually the PR worked fine on my local. However didn't take a closer look on the code part. Will take care from next time

File: app/routes/public/role-invites.js (hunk 1 above)
Comment (niranjan94) [reply app/routes/public/role-invites.js]: @uds5501 you cannot use `await` without using `async` perhaps that's why you had gotten the error.

Comment (niranjan94) [reply app/routes/public/role-invites.js]: But the implementation is still wrong. When using async-await, don't use then-catch. 

Please [read up more on async-await](https://javascript.info/async-await).

**Comment by niranjan94** [comment]
> actually the PR worked fine on my local. However didn't take a closer look on the code part. Will take care from next time

@shreyanshdwivedi Thanks ! Always look at the code itself before running it locally.

**Comment by uds5501** [comment]
@niranjan94 in this commit I have implemented the second loader within another asynchronous function and user `await` withing that instead.
As for using then/catch, if I understood what you wanted correctly, I have used `try/catch` blocks instead for exception handling.

File: app/routes/public/role-invites.js [hunk 2]
```diff
@@ -1,51 +1,59 @@
 import Route from '@ember/routing/route';
 
 export default Route.extend({
-  beforeModel(transition) {
+  async routingModel(payload, user, transition) {
+    try {
+      if (this.authManager.currentUser.email === user.email) {
+        await this.loader
+          .post('/role_invites/accept-invite', payload)
+          .then(invite => {
```
Comment (iamareebjamal) [inline app/routes/public/role-invites.js]: You're using await, then no need of then

Comment (uds5501) [reply app/routes/public/role-invites.js]: @iamareebjamal Okay, I see the error. Fixing it

**Comment by uds5501** [comment]
@iamareebjamal @CosmicCoder96 @niranjan94 Please review!!

File: app/routes/public/role-invites.js [hunk 3]
```diff
@@ -1,51 +1,53 @@
...
-                event       : `${transition.params.public.event_id}`,
-                inviteToken : `${transition.queryParams.token}`,
-                inviteEmail : `${user.email}`
-              }
-            });
-
+    try {
+      let user = await this.loader.post('/role_invites/user', payload);
+      if (this.session.isAuthenticated) {
+        this.routingModel(payload, user);
```
Comment (niranjan94) [inline app/routes/public/role-invites.js]: If routingModel is asynchronous, shouldn't you be awaiting it too ?

**Comment by iamareebjamal** [comment]
Not related to this PR, but use const always unless you need mutable variables

**Comment by uds5501** [comment]
@niranjan94 I need a little help in reducing the repetition.
```js

this.transitionTo('register', {
          queryParams: {
            event       : `${transition.resolvedModels.public.originalId}`,
            inviteToken : `${transition.to.queryParams.token}`,
            inviteEmail : `${user.email}`
          }
        });
```
is the repeating transition which can simply be performed in the `beforeModel`
 IF `this.transitionTo(getRoute, invite.event);` doesn't happen. I tried remodelling the functions to

```js
try {
      if (this.authManager.currentUser.email === user.email) {
        let invite = await this.loader.post('/role_invites/accept-invite', payload);
        let getRoute = (invite.role === 'organiser' || invite.role === 'coorganizer') ? 'events.view' : 'public';
        this.transitionTo(getRoute, invite.event);
        return true
      } else {
        this.set('session.skipRedirectOnInvalidation', true);
        this.session.invalidate();
        return false
      }
    }
```
and wanted to utilise this value to remove this redundant transition but I can't seem to be able to synchronise it.

**Comment by niranjan94** [comment]
@uds5501 would this work ? This is an equivalent of your existing code but without code-repetition.


```js
import Route from '@ember/routing/route';

export default Route.extend({
  async beforeModel(transition) {
    const { token } = transition.to.queryParams;
    const originalEventId = transition.resolvedModels.public.originalId;
    const payload = {
      data: { token }
    };

    const user = await this.loader.post('/role_invites/user', payload);

    if (this.session.isAuthenticated) {

      if (this.authManager.currentUser.email === user.email) {
        return this.transitionTo(
          ['organiser', 'coorganizer'].includes(invite.role) ? 'events.view' : 'public',
          await this.loader.post('/role_invites/accept-invite', payload)
        );
      }

      this.set('session.skipRedirectOnInvalidation', true);
      this.session.invalidate();
    }

    this.transitionTo('register', {
      queryParams: {
        event       : originalEventId,
        inviteToken : token,
        inviteEmail : user.email
      }
    });
  }
});
```

Also, lack of try catch is intentional here. We're letting ember handle the error and render the error page.

**Comment by uds5501** [comment]
> @uds5501 would this work ? This is an equivalent of your existing code but without code-repetition.
> 
> ```js
> import Route from '@ember/routing/route';
> 
> export default Route.extend({
>   async beforeModel(transition) {
>     const { token } = transition.to.queryParams;
>     const originalEventId = transition.resolvedModels.public.originalId;
>     const payload = {
>       data: { token }
>     };
> 
>     const user = await this.loader.post('/role_invites/user', payload);
> 
>     if (this.session.isAuthenticated) {
> 
>       if (this.authManager.currentUser.email === user.email) {
>         return this.transitionTo(
>           ['organiser', 'coorganizer'].includes(invite.role) ? 'events.view' : 'public',
>           await this.loader.post('/role_invites/accept-invite', payload)
>         );
>       }
> 
>       this.set('session.skipRedirectOnInvalidation', true);
>       this.session.invalidate();
>     }
> 
>     this.transitionTo('register', {
>       queryParams: {
>         event       : originalEventId,
>         inviteToken : token,
>         inviteEmail : user.email
>       }
>     });
>   }
> });
> ```
> 
> Also, lack of try catch is intentional here. We're letting ember handle the error and render the error page.

This should work! I will test it in my local and push the changes.

**Comment by uds5501** [comment]
@niranjan94 it just took a minor change otherwise your code worked. Please review!

**Comment by uds5501** [comment]
@niranjan94 @CosmicCoder96 please review

**Comment by uds5501** [comment]
@mariobehling For now, I have made sure that organiser and co-organisers don't get a 400 error when accepting role invites. @niranjan94 @kushthedude @shreyanshdwivedi @mrsaicharan1 , Please review this

**Comment by kushthedude** [comment]
@uds5501 Any GIF or Demo of the following would be awesome

**Comment by uds5501** [comment]
@kushthedude 
ezgif com-video-to-gif (9)

**Comment by kushthedude** [comment]
@uds5501 Looks great, Just one more doubt what if the invite is of other roles, Is the user gets redirected to the event dashboard?

**Comment by uds5501** [comment]
> @uds5501 Looks great, Just one more doubt what if the invite is of other roles, Is the user gets redirected to the event dashboard?

@kushthedude 
For now, let's focus on organisers and co-organisers. For the other roles, we will need to configure permissions individually in the server, I believe.

**Comment by kushthedude** [comment]
> > @uds5501 Looks great, Just one more doubt what if the invite is of other roles, Is the user gets redirected to the event dashboard?
> 
> @kushthedude
> For now, let's focus on organisers and co-organisers. For the other roles, we will need to configure permissions individually in the server, I believe.

If that's the case, I think the PR is good to go then!

**Pull request merged**